Introduce a simplified dependency injection container

Konamiman opened the pull request:
### Changes proposed in this Pull Request:

This pull request introduces a simplified dependency injection container as a replacement of the PHP League's Container package. The motivation for this change is the performance gain, and the fact that we are effectively only using the container for one task: resolving single-instance classes by class name (we don't need a full-blown dependency injection engine for that).

This pull request introduces two new classes:

* `RuntimeContainer`: this class is a replacement for `ExtendedContainer` as the underlying dependency injection container used by WooCommerce at runtime. It doesn't implement explicit class registration, instead it considers any class belonging to the `Automattic\WooCommerce` namespace as implicitly registered as a single-instance class. It uses reflection to resolve (recursively) the dependencies passed as parameters to the `init` method of the resolved classes.
* `TestingContainer`: inheriting from the previous one, includes additional methods to replace the regular classes with mocks and it's intended for usage when running unit tests (being, too, a replacement for `ExtendedContainer`).

`RuntimeContainer` is aware of the containers that the Store API and Blocks implement on their own, and will delegate class resolutions to these containers for classes within the respective namespaces.

Temporarily it's possible to force WooCommerce to still use `ExtendedContainer` at runtime by either creating a `WOOCOMMERCE_USE_OLD_DI_CONTAINER` constant with a value of `true`, or hooking on the `woocommerce_use_old_di_container` filter and having it returning true. This should be done only if issues are found in the new container and the site is broken as a result. `ExtendedContainer` (along with the PHP Lague's container package and all the supporting code) will be removed in WooCommerce 10.0.

Smilarly, it's possible to force WooCommerce to still use `ExtendedContainer` for unit tests by defining the `USE_OLD_DI_CONTAINER` environment variable before triggering the tests run.

### How to test the changes in this Pull Request:

1. Verify that the WooCommerce critical flows (like placing an order from the store and managing products and orders from admin) work with and without adding `define('WOOCOMMERCE_USE_OLD_DI_CONTAINER', true);` in `wp-config.php` (or alternatively, with and without adding this snippet: `add_filter('woocommerce_use_old_di_container', '__return_true');`
2. Verify that all the unit tests pass.
3. Temporarily add the following method to the `Container` class (`src/Container.php`), then run `wp eval 'wc_get_container()->instantiate_everything();'`, again with both the old and the new containers (see step 1):

```php
// Use the current container to retrieve an instance of each of the classes that are registered with the old container.
public function instantiate_everything() {
    foreach ( $this->get_service_providers() as $provider ) {
        $refclass = new \ReflectionClass( $provider );
        $provides_prop = $refclass->getProperty( 'provides' );
        $provides_prop->setAccessible( true );
        $provides_list = $provides_prop->getValue( $refclass->newInstance() );
        foreach ( $provides_list as $class_name ) {
            echo get_class( $this->get( $class_name ) ) . "\n";
        }
    }
}
```

4. Review [the notice added to `src/README.md`](https://github.com/woocommerce/woocommerce/blob/simplify-dependency-injection-container/plugins/woocommerce/src/README.md#important-notice).

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

Konamiman committed:
Add the RuntimeContainer class.

This class is a replacement for ExtendedContainer as the underlying
dependency injection container used in WooCommerce at runtime.
A separate class will be developed as a replacement for unit test run time.

It's possible to force WooCommerce to sill use ExtendedContainer
at runtime by either creating a WOOCOMMERCE_USE_OLD_DI_CONTAINER
constat with a value of 'true', or hooking on the
'woocommerce_use_old_di_container' filter and returning true.

Konamiman committed:
Add changelog file

Konamiman committed:
Merge branch 'trunk' into simplify-dependency-injection-container

Konamiman committed:
Merge branch 'trunk' into simplify-dependency-injection-container

Konamiman committed:
Improve exception message

Konamiman committed:
Add support for classes handled by the Blocks container.

prettyboymp commented on plugins/woocommerce/src/Internal/DependencyManagement/RuntimeContainer.php:
```diff
@@ -0,0 +1,217 @@
+		}
+
+		// Account for the containers used by the Store API and Blocks.
+		if ( StringUtil::starts_with( $class_name, 'Automattic\WooCommerce\StoreApi\\' ) ) {
```
Is this something that the current container implementation already allows?  If not, I think it would be best to avoid allowing the Woo container to obscurely cross over to retrieve a provider from another service container.  Otherwise the packages could get deeply coupled together since cross package dependencies could end up embedded deep in that stack instead of having to be shared at initialization.

Konamiman replied:
There's [one instance where a core class depends on a Store API class](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Internal/Orders/OrderAttributionBlocksController.php#L54). I guess it could be modified so that this particular dependency is explicitly grabbed from the Store API container whenever it's needed (and this might require adjusting some unit tests).

Anyway, the new container is supposed to be able retrieve any class from the `Automattic\WooCommerce` namespace, and it might be a bit weird if we had these two as exceptions (we would have to throw a "please use this other container" error); also the plan is to merge all three container in a single one as the next step of this project, so this explicit redirection to other containers is temporary.

joshuatf replied:
This is also a temporary issue during the transition, right?  Long-term do we plan to consolidate behind a single container, removing the `StoreApi` container?  Or are there specific methods/blockers that exist to doing that?

Konamiman replied:
In principle, the plan is indeed to remove the Store API and the Blocks container, and use this new simplified container for everything. This will probably mean that we'll need to add some additional functionality (e.g. right now the Blocks container is using instantiation callbacks) but the details (and blockers, if any) will surface after a more detailed look at the existing containers.

prettyboymp commented on plugins/woocommerce/src/Internal/DependencyManagement/RuntimeContainer.php:
```diff
@@ -0,0 +1,217 @@
+			return BlocksPackage::container()->get( $class_name );
+		}
+
+		if ( in_array( $class_name, $resolve_chain, true ) ) {
```
Do you have an idea of how big the `$resolve_chain` may get?  If it frequently gets a couple dozen items, it may make sense, for performance, to use a keyed array instead, e.g. `$resolved_set[ $class_name ] = true`. 

- Once the list gets so large, the cost of creating the hash is cheaper than the iterations.
- It doesn't look like the order is needed during the recursion, so a stack isn't necessary.
- All of the classes are going to start with the same prefixes so more of every class name will have to be checked.

Can we at least make `::get_core()` private so we can pivot in the future without the worry of backward compatibility.  Or, even better, make it a final class until we have a reason for it not to be.

Konamiman replied:
> Do you have an idea of how big the $resolve_chain may get?

There are a couple of convoluted cases (related to HPOS) in which the array gets 16 items, but otherwise it never gets more than 5.

> Can we at least make ::get_core() private

I'll probably need `get_core` to be accessible from the inherited `TestingContainer` class I'll develop. But don't worry about this: the class is in the `Internal` namespace so we don't have to keep backwards compatibility.

prettyboymp replied:
> But don't worry about this: the class is in the Internal namespace so we don't have to keep backwards compatibility.

Is that true, though?  

`\Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore` is internal, but a major extension, WooCommerce Subscriptions, extends the class and overrides protected properties and methods.  Making changes to protected interfaces in the `OrdersTableDataStore` class would be a big deal.

I think we need to take extra steps to be certain we have the ability to change underlying implementation in the container class as needed, especially if we're considering merging the container from other packages into it.

Konamiman replied:
This has happened too with [the batch processing engine](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Internal/BatchProcessing/BatchProcessingController.php), but in both cases the "external" code is still managed by us (Automattic/WooCommerce). And while we must be careful with any changes indeed, the point is that to the outside world (extension developers) this code is private for all intents and purposes, it's documented as such, and we have freedom to change it whenever we need to do so.

wavvves replied:
Hi @Konamiman, great initiative first of all! I'm reviewing the changes, but before that I just wanted to say I'm a bit with @prettyboymp here, what third-party shouldn't do but can technically do has not proven to be a reliable binding contract :/
Especially with access modifiers, not getting them locked in to start with might bite us in the future, and prevent changes in the future without breaking stores. Ultimately it is third-party responsibility to keep up with us, but that usually ends with a code freeze exception on our end :(

Konamiman replied:
We'll do a public announcement about this, including the mehcanism to switch to the old container and its removal in WooCommerce 10.0, right after this is merged. This should be enough, we've done a similar thing with the removal of the legacy REST API (which _was_ an actual breaking change) and it went well.

As for our canonical extensions, see [this comment](https://github.com/woocommerce/woocommerce/pull/52296#discussion_r1836093576).

Konamiman committed:
Introduce the TestingContainer class.

This is a simplified version of the old ExtendedContainer.
When unit tests run an instance of this class will replace the
instance of RuntimeContainer used internally by Container.

If the environment variable USE_OLD_DI_CONTAINER is set
then ExtendedContainer will be used for unit tests instead
(this mechanism will be removed in WooCommerce 10.0).

Konamiman committed:
Add unit tests for TestingContainer

Konamiman committed:
Fix linting issues

Luc45 commented:
> The motivation for this change is the performance gain

Just curious: DI containers are usually [pretty fast](https://kocsismate.github.io/php-di-container-benchmarks/benchmark.html). Is there a profiling (like Xdebug profiler) that indicates a performance issue with the current DI implementation?

prettyboymp commented:
> > The motivation for this change is the performance gain
> 
> Just curious: DI containers are usually [pretty fast](https://kocsismate.github.io/php-di-container-benchmarks/benchmark.html). Is there a profiling (like Xdebug profiler) that indicates a performance issue with the current DI implementation?

@Luc45  https://github.com/woocommerce/woocommerce/issues/50714#issuecomment-2297480640

The current DI implementation iterates through implementations to find the matching dependency implementation and doesn't implement any sort of caching/memoization.

prettyboymp commented on plugins/woocommerce/tests/Tools/TestingContainer.php:
```diff
@@ -0,0 +1,91 @@
+/**
+ * Dependency injection container to be used in unit tests.
+ */
+class TestingContainer extends RuntimeContainer {
```
Instead of extending the RuntimeContainer for testing purposes, can't the TestingContainer be a composition wrapping the real container, especially since we're just shortcutting mappings with the $replacements?  This would avoid the need to potentially expose the container to extension.

Konamiman replied:
This would make things more complex. We need to access the runtime container's cache in order to inject the replacement of `LegacyProxy` with `RuntimeLegacyProxy`, so either that property needs to be protected or we need to resort to reflection to get its value. Additionally, we would need to instantiate a new runtime container within each call to `reset_all_resolved`.

Again, `RuntimeContainer` is in `Internal` so private for all practical purposes, we created that namespace precisely for these cases. If we need to emphasize this more prominently in the documentation, we'll do that.

prettyboymp replied:
Emphasizing it more and making sure that A8C/WooCommerce developed extensions don't exist as examples ignoring fact that this should be considered internal is something we should improve upon.

Konamiman replied:
I have downloaded and unzipped all the WooCommerce canonical extensions (p6q6po-e59-p2), then did the following searches for possible usages of internal code related to this pull request:

* `Automattic\WooCommerce\Internal\DependencyManagement\ExtendedContainer`

No matches were found. Note that WooCommerce Payments does have an `ExtendedContainer` class, but it's a custom implementeation that's **not** tied to the one in WooCommerce core.

* `League\Container`

Found this in WooCommerce Payments and Google for WooCommerce, but again, these are custom copies of the package (with a custom namespace) and they do **not** use any of the code in the package contained in WooCommerce in any way.

I have also created a Jurassic Ninja site with all the extensions installed and active, did some testing (including placing an order and paying with credit card, so WooCommerce Payments is used) and there were no issues. Additionally, I physically removed the core `ExtendedContainer` and `League\Container` classes (which is what we would do in WooCommerce 10.0) and everything continued working (and _then_ if I add `define('WOOCOMMERCE_USE_OLD_DI_CONTAINER', true);` I get a fatal error, as expected - this was just to make sure that I had actually and effectively removed the classes).

Therefore, as far as canonical extensions are concerned, it seems that we're safe with this change.

Luc45 commented:
> The current DI implementation iterates through implementations to find the matching dependency implementation and doesn't implement any sort of caching/memoization.

Got it, thanks for the clarification. I've done some profiling myself and I was going to say that Blocks are the real time hog, but I see this is already being tackled by another team. Excited for these performance improvements! :rocket: 

> It uses reflection to resolve (recursively) the dependencies passed as parameters to the init method of the resolved classes.

Love this, so Woo DI will have autowiring?

Konamiman committed:
Do the recursive resolution detection earlier in the process

Konamiman commented:
> Love this, so Woo DI will have autowiring?

Yes, autowiring will be in fact the only available mechanism in the new container; even with the usage of reflection it gets faster than the old container.

> Is there a profiling (like Xdebug profiler) that indicates a performance issue with the current DI implementation?

We haven't published any figures, but some profiling I performed shows that with the new container [the `init_hooks` method](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/class-woocommerce.php#L284) takes 1/3 of the time it takes with the old one (the time used by the code in the method before the container usage is negligible).

Konamiman committed:
Update src/README.md

Konamiman committed:
Lint the README

joshuatf commented on plugins/woocommerce/src/Internal/DependencyManagement/RuntimeContainer.php:
```diff
@@ -0,0 +1,225 @@
+	 *
+	 * @return bool True if this class should be used as the core WooCommerce dependency injection container, false if ExtendedContainer should be used instead.
+	 */
+	public static function must_use(): bool {
```
Nitpick, could we rename this to `should_use`?  This matches the above comment semantics.

joshuatf commented on plugins/woocommerce/src/README.md:
```diff
@@ -43,36 +40,50 @@ If you don't have Composer installed, go and check how to [install Composer](htt
+1. `define('WOOCOMMERCE_USE_OLD_DI_CONTAINER', true);`
+2. `add_filter('woocommerce_use_old_di_container', '__return_true');`
+
+This should be done **only** if an issue in the new container causes disruption in the site. This fallback mechanism, together with the old container and the PHP League's Container package, will be removed in WooCommerce 10.0.
```
This is probably implied, but should we explicitly ask users to open issues they encounter with the new container so we can get them fixed by 10.0?

joshuatf commented on plugins/woocommerce/src/README.md:
```diff
@@ -43,36 +40,50 @@ If you don't have Composer installed, go and check how to [install Composer](htt
+
+This should be done **only** if an issue in the new container causes disruption in the site. This fallback mechanism, together with the old container and the PHP League's Container package, will be removed in WooCommerce 10.0.
+
+The new container is used in the same way as the old one: it's only for classes in the `src` directory, dependencies are defined in an `init` method of the classes, the container exposes a `get` method and a `has` method, and `wc_get_container` can be used to use the container from legacy code; the difference is that explicit class registration is not needed anymore (any class in the `Automattic\ ...
```
Should we link to the [other docs](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/README.md/#L6-L7) for this?

joshuatf commented on plugins/woocommerce/src/Internal/DependencyManagement/RuntimeContainer.php:
```diff
@@ -0,0 +1,225 @@
+		}
+
+		if ( ! $this->is_class_allowed( $class_name ) ) {
+			throw new ContainerException( "Attempt to get an instance of class '$class_name', which is not in the " . self::WOOCOMMERCE_NAMESPACE . ' namespace. Did you forget to add a namespace import?' );
```
Nice check here!

joshuatf reviewed (commented):
Thanks for taking the time to work on this, @Konamiman!  Overall this is testing well for me.

* ✅ Generally smoketested WooCommerce, including admin and product admin screens
* ✅ Unit tests are passing
* ✅ Checked that the legacy container works with `WOOCOMMERCE_USE_OLD_DI_CONTAINER` defined
* ✅ Validated the loaded dependencies in both scenarios with `instantiate_everything()`

Left a few nitpicks/comments in the code, but didn't find any glaring issues.  Definitely good to get more eyes on this before merging since it touches so much of the application.

prettyboymp commented on plugins/woocommerce/includes/class-woocommerce.php:
```diff
@@ -335,14 +335,13 @@ private function init_hooks() {
+		$container->get( Automattic\WooCommerce\Internal\Orders\OrderAttributionController::class )->register();
+		$container->get( Automattic\WooCommerce\Internal\Orders\OrderAttributionBlocksController::class )->register();
+		$container->get( Automattic\WooCommerce\Internal\CostOfGoodsSold\CostOfGoodsSoldController::class )->register();
+		$container->get( Automattic\WooCommerce\Internal\Utilities\LegacyRestApiStub::class )->register();    }
```
Looks like the lint rules missed this closing bracket on the previous line.

Konamiman replied:
Really weird how the CI linting job isn't complaining about this one! 😮 I'll fix it, thanks.

coreymckrill commented on plugins/woocommerce/src/Internal/DependencyManagement/RuntimeContainer.php:
```diff
@@ -0,0 +1,225 @@
+	 *
+	 * By default, this returns true, to have it return false you can:
+	 *
+	 * 1. Define the WOOCOMMERCE_USE_OLD_DI_CONTAINER constant with a value of true; or
+	 * 2. Hook on the 'woocommerce_use_old_di_container' filter and have it return false (it receives the value of WOOCOMMERCE_USE_OLD_DI_CONTAINER, or false if the constant doesn't exist).
```
🤔 Would it be worth adding something to the System Status Report showing whether the old DI container is being used? And/or adding something to WC Tracker? That way we could get a sense for how disruptive it might be to remove the old DI container in versionn 10.0.

Konamiman replied:
This makes sense, but I'd add this in a separate PR, after we discuss the approach to follow.

Konamiman committed:
Rename 'must_use' to 'should_use', and tweak the README

xristos3490 commented on plugins/woocommerce/src/Internal/DependencyManagement/RuntimeContainer.php:
```diff
@@ -0,0 +1,225 @@
+		}
+
+		$init_method = $ref_class->getMethod( 'init' );
+		if ( ! $init_method->isPublic() || $init_method->isStatic() ) {
```
Should we require only `final` injection methods, as before? 🤔

Konamiman replied:
I think having the check in the linter is enough, but I'm open to discussion if there's any tangible benefit on enforcing `final`-ness at runtime.

xristos3490 replied:
Honestly, I don’t see much benefit to enforcing `final` beyond consistency. The linter should handle it, so I’m good with keeping it as-is if that works for everyone.

xristos3490 commented on plugins/woocommerce/tests/legacy/bootstrap.php:
```diff
@@ -172,26 +180,33 @@ private function initialize_hpos() {
+		$container       = wc_get_container();
+		$inner_container = $inner_container_property->getValue( $container );
+		if ( $use_old_container ) {
+			$inner_container->replace( LegacyProxy::class, MockableLegacyProxy::class );
```
I don't think this is a blocker, but when I run tests with the `USE_OLD_DI_CONTAINER` the system fails on my end.

Although the `$use_old_container` is set correctly, the inner container appears to be instantiated as RuntimeContainer. Maybe a race condition?


The command:
`USE_OLD_DI_CONTAINER=YES pnpm test:php`

The error:
```
Error in bootstrap script: Error:
Call to undefined method Automattic\WooCommerce\Internal\DependencyManagement\RuntimeContainer::replace()
#0 /Users/chris/Repos/woocommerce/plugins/woocommerce/tests/legacy/bootstrap.php(96): WC_Unit_Tests_Bootstrap->initialize_dependency_injection(true)
#1 /Users/chris/Repos/woocommerce/plugins/woocommerce/tests/legacy/bootstrap.php(329): WC_Unit_Tests_Bootstrap->__construct()
#2 /Users/chris/Repos/woocommerce/plugins/woocommerce/tests/legacy/bootstrap.php(336): WC_Unit_Tests_Bootstrap::instance()
#3 /Users/chris/Repos/woocommerce/plugins/woocommerce/vendor/phpunit/phpunit/src/Util/FileLoader.php(66): include_once('/Users/chris/Re...')
#4 /Users/chris/Repos/woocommerce/plugins/woocommerce/vendor/phpunit/phpunit/src/Util/FileLoader.php(49): PHPUnit\Util\FileLoader::load('/Users/chris/Re...')
#5 /Users/chris/Repos/woocommerce/plugins/woocommerce/vendor/phpunit/phpunit/src/TextUI/Command.php(567): PHPUnit\Util\FileLoader::checkAndLoad('./tests/legacy/...')
#6 /Users/chris/Repos/woocommerce/plugins/woocommerce/vendor/phpunit/phpunit/src/TextUI/Command.php(347): PHPUnit\TextUI\Command->handleBootstrap('./tests/legacy/...')
#7 /Users/chris/Repos/woocommerce/plugins/woocommerce/vendor/phpunit/phpunit/src/TextUI/Command.php(114): PHPUnit\TextUI\Command->handleArguments(Array)
#8 /Users/chris/Repos/woocommerce/plugins/woocommerce/vendor/phpunit/phpunit/src/TextUI/Command.php(99): PHPUnit\TextUI\Command->run(Array, true)
#9 phpvfscomposer:///Users/chris/Repos/woocommerce/plugins/woocommerce/vendor/phpunit/phpunit/phpunit(106): PHPUnit\TextUI\Command::main()
#10 /Users/chris/Repos/woocommerce/plugins/woocommerce/vendor/bin/phpunit(118): include('phpvfscomposer:...')
```

Konamiman replied:
I can't reproduce this but it looks like a race condition indeed. In `WC_Unit_Tests_Bootstrap::__construct` the existence of the `USE_OLD_DI_CONTAINER` variable is checked, if found then the `WOOCOMMERCE_USE_OLD_DI_CONTAINER` constant is defined; looks like by then the main container (`src/Container.php`) has been instantiated already and thus it's using a `RuntimeContainer`. But I don't understand why this is not happening for me (and others who have already tested). Is there anything "unusual" in your setup?

xristos3490 replied:
Thanks for looking into this @Konamiman!

> Is there anything "unusual" in your setup?

I realized that I had hardcoded the `should_use` value during testing. After reverting this change, everything works as expected. 😅

Konamiman replied:
Well that was unusual, for sure 😆 (I can't laugh too hard, though, as I have done similar things in the past)

xristos3490 reviewed (commented):
Nice work, @Konamiman!

✔️  Did some exploratory testing, and everything is looking good on my end. 
✔️ Tested the before and after classes using the `instantiate_everything` method. 
✔️  Verified that the `WOOCOMMERCE_USE_OLD_DI_CONTAINER` constant works as intended.
❓ ~Ran the unit tests with the `USE_OLD_DI_CONTAINER` environment variable but encountered an error. Please refer to the comment below for details. Not sure if it’s a blocker.~
✔️ **Update:** Unit tests work as expected.

roykho commented:
✅ Initial smoke test went well. Nothing is broken.
✅ Tested with and without `add_filter('woocommerce_use_old_di_container', '__return_true');`
✅ Tested with `instantiate_everything` drop in code with new/old DI container.

Konamiman committed:
Merge branch 'trunk' into simplify-dependency-injection-container

Konamiman committed:
Small fix to README

lanej0 reviewed (commented):
I ran the core end to end suite against the different testing scenarios outlined, and didn't see any issues.  Looks good from the Solaris perspective!

jaclync commented:
The Woo mobile team smoke-tested the apps and we didn't find any issues 👍 (Ref: p1730220985608679-slack-C6H8C3G23)

adrianduffell commented:
✅ Smoke-tested coming soon mode and found no issues.

> The motivation for this change is the performance gain, and the fact that we are effectively only using the container for one task: resolving single-instance classes by class name (we don't need a full-blown dependency injection engine for that).

Raising a concern that this could limit us from using more powerful DI patterns in the future. Is it possible to please elaborate on the performance issues vs this tradeoff?

wavvves commented on plugins/woocommerce/src/Internal/DependencyManagement/ExtendedContainer.php:
```diff
@@ -15,6 +15,8 @@
  * This class extends the original League's Container object by adding some functionality
  * that we need for WooCommerce.
+ *
+ * NOTE: This class will be removed in WooCommerce 10.0.
```
This might have limited visibility to any third-party extending or modifying it, can we make use of  the @deprecated tag and maybe adding a doing_it_wrong() so we can track edge cases that might break a store once released?

Also, was it considered shipping this as disabled experimental, with an option to enable on WC > Settings > features?

Konamiman replied:
> This might have limited visibility to any third-party extending or modifying it

This class is inside the `Internal` namespace, this means that [nobody outside we the core maintainers is expected to be using it](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Internal/README.md). The ability to still use it is provided only as a temporary fallback mechanism in case issues are found in the new container. This comment is more a reminder for ourselves.

> was it considered shipping this as disabled experimental, with an option to enable on WC > Settings > features?

This is not really a "feature" in active development but an atomic change in how we organize things internally. Providing it as a feature that can be toggled via UI (and worse, disabled by default) could give the false impression that this is an optional thing and slow down adoption.

Konamiman replied:
> nobody outside we the core maintainers is expected to be using it

That said: it's probably a good idea to post a developer advisory about this in the Woo developers blog.

wavvves reviewed (commented):
✅ Verified that the mentioned flows work with and without define('WOOCOMMERCE_USE_OLD_DI_CONTAINER', true);.
✅ Verified that all unit tests pass.
✅ Successfully ran wp eval `wc_get_container()->instantiate_everything();` with both old and new containers after adding the method to Container.php.
✅ Reviewed the notice added to src/README.md.

Konamiman commented:
> Raising a concern that this could limit us from using more powerful DI patterns in the future.

We introduced the container [four years ago now](https://github.com/woocommerce/woocommerce/pull/27733) and so far we haven't found the need to use the container in a more sophisticated way. So I think it's better to stick to [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) and do whatever is beneficial now, we can always change our minds and move to a more powerful approach in the future (we can do that because all the affected code is in [the `Internal` namespace](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Internal/README.md)).

> Is it possible to please elaborate on the performance issues vs this tradeoff?

See p7bje6-6bl-p2. Or for a TLDR: https://github.com/woocommerce/woocommerce/pull/52296#issuecomment-2443761709

albarin reviewed (approved):
🟢 Smoke tested and everything looks good (both with and without the `WOOCOMMERCE_USE_OLD_DI_CONTAINER` constant).
🟢 All unit tests passing.
🟢 Tested the `wp eval 'wc_get_container()->instantiate_everything();'`.

Approving! TY!
